### PR TITLE
check and warn about Scio version #335

### PR DESCRIPTION
**Pull request by nevillelyh**
This is a run-time approach. Compile-time approach isn't hard but would require some form of macro that I still wanna avoid for now.

### CODE REVIEW

**Comment by nevillelyh** [comment]
@ravwojdyla thoughts?

File: scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import com.google.api.client.json.jackson2.JacksonFactory
+import org.slf4j.LoggerFactory
+
+import scala.collection.mutable
+import scala.util.Try
+
+private object VersionUtil {
+
+  case class SemVer(major: Int, minor: Int, rev: Int, snapshot: Int) {
+    def toInt: Int = major * 1000000 + minor * 10000 + rev * 100 + snapshot
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: I understand how practically this is gonna work - but in theory this breaks. Counterargument:

```
my-version: 0.1.101 (0 + 10000 + 10100 = 20100)
release: 0.2.0 (0 + 20000 + 0 = 20000)
```

I think you could use Algebird's `Min`:

```
import com.twitter.algebird._
import com.twitter.algebird.Operators._
Min((0,2,0,"")) + Min((0,1,101,"")) -> Min((0,1,101,))
Min((0,2,0,"")) + Min((0,2,0,"SNAPSHOT")) -> Min((0,2,0,))
```

File: scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+  private def parseVersion(version: String): SemVer = {
+    val m = pattern.findFirstMatchIn(version).get
+    // higher value for no "-SNAPSHOT"
+    val snapshot = if (m.group(4) == "-SNAPSHOT") 0 else 1
+    SemVer(m.group(1).toInt, m.group(2).toInt, m.group(3).toInt, snapshot)
+  }
+
+  def checkVersion(current: String, latest: Option[String]): Seq[String] = {
+    val b = mutable.Buffer.empty[String]
```
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: You could probably just `++` two `Seq` - and skip the mutable buffer?

File: scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+private object VersionUtil {
+
+  case class SemVer(major: Int, minor: Int, rev: Int, snapshot: Int) {
+    def toInt: Int = major * 1000000 + minor * 10000 + rev * 100 + snapshot
+  }
+
+  private val TIMEOUT = 3000
+  private val url = "https://api.github.com/repos/spotify/scio/releases"
+  private val pattern = """v?(\d+)\.(\d+).(\d+)(-SNAPSHOT)?""".r
+  private val logger = LoggerFactory.getLogger("VersionUtil")
```
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: Fixed

File: scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala (hunk 1 above)
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: That'd require a tuple, not case class and Algebird dependency. Also we'd never hit 100 in any fields. Would prefer to keep it simple here.

File: scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala (hunk 2 above)
Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: Yes but that'd require 2 `if () Seq("...") else Nil` blocks which IMO is slightly more cryptic.

File: scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala (hunk 1 above)
Comment (ravwojdyla) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: case class + no algebird:
```
case class Ver(major: Int, minor: Int, rev: Int, classifier: String) extends Ordered[Ver] {
  def compare(that: Ver): Int = {
    implicit val revStringOrder = Ordering[String].reverse
    implicitly[Ordering[(Int, Int, Int, String)]]
      .compare(Ver.unapply(this).get, Ver.unapply(that).get)
  }
}
```

Comment (nevillelyh) [inline scio-core/src/main/scala/com/spotify/scio/VersionUtil.scala]: You win :)

**Pull request merged**